testing/internal/e2e/tests/static/target_address_test.go

```diff
@@ -0,0 +1,91 @@
+package static_test
+
+import (
+	"context"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/boundary/testing/internal/e2e"
+	"github.com/hashicorp/boundary/testing/internal/e2e/boundary"
+	"github.com/stretchr/testify/require"
+)
+
+// TestTargetAddress uses the boundary cli to create a number of supporting objects
+// to connect to a target. It then attempts to connect to that target using a host source
+// and verifies that the connection was successful. Afterwards, we validate that a address
+// cannot be assigned directly to the target that has an existing host source association.
+// The host sources are removed from the same target and a network address is assigned
+// directly to the target. Finally, it then attempts to connect to that target using
+// the direct network address association.
+func TestTargetAddress(t *testing.T) {
+	e2e.MaybeSkipTest(t)
+	c, err := loadConfig()
+	require.NoError(t, err)
+
+	ctx := context.Background()
+	boundary.AuthenticateAdminCli(t, ctx)
+	newOrgId := boundary.CreateNewOrgCli(t, ctx)
+	t.Cleanup(func() {
+		ctx := context.Background()
+		boundary.AuthenticateAdminCli(t, ctx)
+		output := e2e.RunCommand(ctx, "boundary", e2e.WithArgs("scopes", "delete", "-id", newOrgId))
+		require.NoError(t, output.Err, string(output.Stderr))
+	})
+	newProjectId := boundary.CreateNewProjectCli(t, ctx, newOrgId)
+	newHostCatalogId := boundary.CreateNewHostCatalogCli(t, ctx, newProjectId)
+	newHostSetId := boundary.CreateNewHostSetCli(t, ctx, newHostCatalogId)
+	newHostId := boundary.CreateNewHostCli(t, ctx, newHostCatalogId, c.TargetIp)
+	boundary.AddHostToHostSetCli(t, ctx, newHostSetId, newHostId)
+	newTargetId := boundary.CreateNewTargetCli(t, ctx, newProjectId, c.TargetPort)
+	boundary.AddHostSourceToTargetCli(t, ctx, newTargetId, newHostSetId)
+
+	// Connect to target and print host's IP address
+	output := e2e.RunCommand(ctx, "boundary",
+		e2e.WithArgs(
+			"connect",
+			"-target-id", newTargetId,
+			"-exec", "/usr/bin/ssh", "--",
+			"-l", c.TargetSshUser,
+			"-i", c.TargetSshKeyPath,
+			"-o", "UserKnownHostsFile=/dev/null",
+			"-o", "StrictHostKeyChecking=no",
+			"-o", "IdentitiesOnly=yes", // forces the use of the provided key
+			"-p", "{{boundary.port}}", // this is provided by boundary
+			"{{boundary.ip}}",
+			"hostname", "-i",
+		),
+	)
+	require.NoError(t, output.Err, string(output.Stderr))
+
+	parts := strings.Fields(string(output.Stdout))
+	hostIp := parts[len(parts)-1]
+	require.Equal(t, c.TargetIp, hostIp, "SSH session did not return expected output")
+	t.Log("Successfully connected to target")
+
+	_, err = boundary.SetTargetAddressCli(t, ctx, newTargetId, c.TargetIp)
+	require.Error(t, err, "Target update did not return expected error")
```

Comment:
We should probably assert that this error is what we expect (we could string match part of it or something like that)

Reply:
If we want to do any sort of asserting about the content of this error, I think it should be checking what the status code was. The string text could change but the status code should be stable, and should be what customers use. @moduli can correct me but I think we have some utility for parsing out the status code from the error already.

Reply:
I agree. We should validate the status code here. You can parse errors using `boundary.CliError`.

Example:
```
var response boundary.CliError
err = json.Unmarshal(output.Stderr, &response)
require.NoError(t, err)
require.Equal(t, response.Status, 404, "Expected to error when connecting to a target with zero hosts")
```

Reply:
That sounds like a good idea, I will update the test to make checks based on the status code instead of error message.